### ISSUE TITLE
`kubectl run` doesn’t work for simple "run command, parse result"

### ISSUE THREAD

**Issue by janetkuo** [kubernetes/kubernetes#28695]
(source: user feedback)

First, it outputs `Waiting for pod ...` which gets mingled with commands output.  Second, simple output often gets lost.

**1. -it, date**: doesn't work

``` console
$ kubectl run test --rm --restart=Never -it --image=ubuntu -- date
Waiting for pod default/test to be running, status is Pending, pod ready: false
Error from server: Unrecognized input header
```

**2. -it, bash sleep, then date**: works with output mingled with command output 

``` console
$ kubectl run test --rm --restart=Never -it --image=ubuntu -- bash -c "sleep 3; date; sleep 3"
Waiting for pod default/test to be running, status is Pending, pod ready: false

Hit enter for command prompt
                            Fri Jul  8 17:19:50 UTC 2016
pod "test" deleted
```

**3. -t, bash sleep, then date**: doesn't work 

``` console
$ kubectl run test --rm --restart=Never -t --image=ubuntu -- bash -c "sleep 3; date; sleep 3"
error: -i/--stdin is required for containers with -t/--tty=true
See 'kubectl run -h' for help and examples.
```

**4. -i, date**: works but output is still mingled 

``` console
$ kubectl run test --rm --restart=Never -i --image=ubuntu -- date
Waiting for pod default/test to be running, status is Pending, pod ready: false
Fri Jul  8 17:21:59 UTC 2016
pod "test" deleted
```

Compared to `docker run`, all 4 cases return the same simple output:

``` console
$ docker run -i ubuntu date
Fri Jul  8 17:23:45 UTC 2016
```

@kubernetes/kubectl

**Comment by janetkuo** [comment]
The `Waiting for pod default/test to be running, status is Pending, pod ready: false` might be crucial for new comers to know what they're waiting for if the pod takes a while to start. We can add a flag to hide this message (I prefer this), or we can just remove it.

**Comment by janetkuo** [comment]
Another issue for case 1 is that we should respect `--rm` and delete the pod even when error happens.

**Comment by thockin** [comment]
Adding a --silent or something might be enough.

**Comment by duglin** [comment]
By default we should assume people don't want their output mucked with.

Only exception I can see to that is if the image needs to be downloaded, but once cached it should be speedy.  So, I'd prefer:
- no output except for "downloading image"
- add a --silent/--quite flag to hide the "downloading image" msg

**Comment by AdoHe** [comment]
```
We can add a flag to hide this message
```
I prefer this too.

**Comment by smarterclayton** [comment]
The attach race logic absolutely needs to get fixed.  I believe that's a
gap in both the kubelet launch and client invocation scenarios.  Or we need
to limit attach to stdin and fix logging to separate stdout and stderr.

**Comment by janetkuo** [comment]
Adding `--quiet` in #28801. 

Some more thoughts from in-person discussion with @bgrant0607 : 

1. We can default the `--quiet` mode when `-it` are both specified, or 
2. Add a flag (e.g. `--interactive`) that'd be equivalent to `--restart=Never --quiet --rm` (depending on what default makes sense for the users)

**Comment by thockin** [comment]
I like the idea of --quiet being implied with -i -t.

I also like the idea of an umbrella flag that covers it all.  --interactive
is the long-form of -i, though, isn't it?  How about --attached or
--one-shot?  Maybe --interactive is better.

Is there a use for -i that isn't this?  Maybe just -i -t together implies
everything?

**Comment by janetkuo** [comment]
@thockin we don't actually have `--interactive` flag in the system. The long-form for `-i` is `--stdin` (although we do mention `-i/--interactive` in kubectl run help by mistake, which is misleading).

**Comment by thockin** [comment]
SGTM, if a bit confusing (to people who think they know what's going on but
seem to be wrong - me :)

**Comment by ardnaxelarak** [comment]
#investigating

**Comment by pwittrock** [comment]
@ardnaxelarak should I assign this to you?

**Comment by ardnaxelarak** [comment]
@pwittrock yes, #dibs

**Comment by MHBauer** [comment]
Sat with @ardnaxelarak and tried to debug case (1) for a few hours.

For this case, the `date` command executes and finishes during one of the pod waiting loops in `pkg/kubectl/cmd/run.go#waitForPodRunning()`.
WaitForPodRunning returns status.Succeeded to handleAttachPod() which decides to do a `Logs` operation on the pod, as there is no place to attach.
Somehow (we're not sure of the details, and are not sure how apiserver communicates to kubelet) we end up in `pkg/kublet/dockertools/kube_docker_client.go#Logs()` where the call to docker using the client with `ContainerLogs()` is made. After, we call `redirectResponseToOutptStream()`. Inside of that, it says that it is not a tty, so we execute the docker code `StdCopy()`, which demultiplexes stdout and stderr from the dock response to the request for logs. Unfortunately, when we created the container, we told docker it was a rawTTY, so docker is returning non-multiplexed output.

The tty setting needs to be pulled from the container config. In `pkg/kubelet/dockertools/docker_manager.go#GetContainerLogs()` it is always set to false.`
```
	sopts := StreamOptions{
		OutputStream: stdout,
		ErrorStream:  stderr,
		RawTerminal:  false,
	}
```
`RawTerminal` needs to be set based on the container config, and not set unconditionally.

We could use a clue as to how to go about achieving this.

**Comment by mikebrow** [comment]
Did some digging around tonight it looks like a polling design is used to finding out if/when the container is running, and the implementation has quite a few sleeps in it. One 2second sleep is hard coded in pkg/kubectl/cmd/run.go waitForPodRunning(). The other sleeps are I believe on the kublet side in the prober where there is an initialDelaySeconds (default 2 for v1) and periodSeconds (default 4 for v1). So no surprise the running state appears to take quite a while to register after the initial run command.  Cheers.

**Comment by brendandburns** [comment]
fwiw, I don't think we should imply `--quiet` with -`it` plenty of newbie users are going to want interactive terminal sessions, but it takes a long-ish time to run for the first time, especially if the image is large, I think that we still want to provide feedback by default.

Definitely makes sense to have a `--quiet` flag though.

**Comment by duglin** [comment]
showing something to indicate that the image is being downloaded makes sense to me (like Docker does), but I believe the current message shows up even when that's not the case and I think that's less than ideal.  Once the image is cached I think our goal should be that: `xxx` and `kubectl run xxx` show the exact same output.

**Comment by pwittrock** [comment]
@ardnaxelarak @brendandburns Should we close this and create another issue for remaining items?  Would printing a message about the --quiet flag when running without --quiet be sufficient?

**Comment by duglin** [comment]
I'd prefer if we didn't make it even more unnecessarily verbose than it already is :-)

**Comment by MHBauer** [comment]
3 is the only case left that requires work or a decision to be made about whether it is a valid use case.

The real decision is about whether to make `--quiet` the default.

There is a third decision about whether to output the extra status info as stdout vs stderr. For reference, in docker, the equivalent `pulling image` text goes to stdout, not stderr. This might split the difference on the `--quiet` as default decision.

```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.973+5a0c22e09ab769-dirty", GitCommit:"5a0c22e09ab7695727b9c1be062f39e346960e9f", GitTreeState:"dirty", BuildDate:"2016-10-26T17:36:07Z", GoVersion:"go1.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.973+5a0c22e09ab769-dirty", GitCommit:"5a0c22e09ab7695727b9c1be062f39e346960e9f", GitTreeState:"dirty", BuildDate:"2016-10-27T00:38:12Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
```
 
`1`. fixed by #31558 
```
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -it --image=ubuntu -- date
Waiting for pod default/test to be running, status is Pending, pod ready: false
Thu Oct 27 00:58:05 UTC 2016
pod "test" deleted
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -it --image=ubuntu --quiet -- date
Thu Oct 27 00:58:15 UTC 2016
```
`2`. correct behavior with `--quiet`
```
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -it --image=ubuntu -- bash -c "sleep 3; date; sleep 3"
Waiting for pod default/test to be running, status is Pending, pod ready: false
If you don't see a command prompt, try pressing enter.
                                                      Thu Oct 27 01:06:20 UTC 2016
Waiting for pod default/test to terminate, status is Running
pod "test" deleted
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -it --image=ubuntu --quiet -- bash -c "sleep 3; date; sleep 3"
Thu Oct 27 01:06:32 UTC 2016
```
`3`. no change due to differing definition of flags from docker. I am not sure this is a valid use case for k8s.
`4`. 
```
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -i --image=ubuntu -- date
Waiting for pod default/test to be running, status is Pending, pod ready: false
Thu Oct 27 01:09:15 UTC 2016
pod "test" deleted
root@masterbuilder:~/go/src/k8s.io/kubernetes# kubectl run test --rm --restart=Never -i --image=ubuntu --quiet -- date
Thu Oct 27 01:09:24 UTC 2016
```

**Comment by duglin** [comment]
My perspective:
- I think we should differentiate between a delay due to pulling an image and a delay due to the time it takes to create the container/pod
- I think the default should be to hide the "waiting for pod" messages. Personally, I would like to get rid of it all together, not even control it by a flag.  If it takes so long for things to start then we should fix that delay instead of be chatty.
- by default I'm ok with what Docker does and shows some kind of "pulling image" message - and then `--quiet` can hide that.
- I'd like to get `kubectl run` to the point where I can use it the same way I do `docker run`, which means use it in scripts and Makefiles. In those cases any extra output will interfere with my expected output and I think the default should be to try to match what the user expects to see from their cmd, not from the container engine.
- This also means I'd prefer if  `--restart=Never` the default too.  I really like that when I convert a normal cmd into a dockerized cmd I just wrapper it with something like `docker run ubuntu`, maybe add `-ti`.  With k8s I need to add: `kubectl run test --restart=Never --image=ubuntu --quiet -- `.  That's lot more IMO and hurts our UX.
- so: `kubectl run test --image=ubuntu [cmd]` would be nice for now.
- ultimately, I don't really like the idea of a require flag (--image), if something is required just force it to be in the args.  Also, I'm not sure why we require a name on 'run' since its often used for one-off containers.  Just let kubectl pick a random name.  My goal: `kubectl run ubuntu [cmd]`.  Wouldn't mind if `-ti` were on by default, not sure why Docker doesn't do that.

**Comment by thockin** [comment]
We can maybe make kubectl close the loop and show more events, but it is a
little different because it is decoupled from the actual pull.

I agree that "waiting" is not that helpful

I agree on --quiet

--restart=Never can't be default without a long transition plan.  We could
introduce `kubectl deploy` as restart=Always, and maybe `kubectl job`?  But
that's kind of ugly.  I think this one is sort of a lost cause.

--image as a flag is nice because it is not positional.  I can not tell you
how many times I have messed up the order in docker run, with a very
unhelpful error message.  I agree that name could be defaulted.

**Comment by bgrant0607** [comment]
@ardnaxelarak Are you still working on this?

**Comment by ardnaxelarak** [comment]
I am not; @MHBauer might be.

**Comment by ncdc** [comment]
I'm not sure if it's mentioned elsewhere, but when we switched from go-dockerclient to engine-api, we lost the ability to retrieve logs generated by the container prior to the time the attaching occurs. I added this functionality back to docker's api, but we'll need to bump the godep (and switch back from engine-api to docker, as the code was reintegrated into the main repo) and then add back support for retrieving logs at the start of attach. This is one reason why when you `kubectl run` you may miss the beginning of the container's output.

**Comment by duglin** [comment]
@ncdc when was that change made?

**Comment by ncdc** [comment]
@duglin we switched to engine-api back around kube 1.3 I think. This is the docker change we'll need: https://github.com/docker/docker/pull/26718.

**Comment by dims** [comment]
@ncdc, @duglin : i believe there's some chatter over engine-api -> docker over at https://github.com/kubernetes/kubernetes/issues/36011

**Comment by cmluciano** [comment]
With @duglin 's PR, the only output one receives is 

```
./kubectl run test --rm --restart=Never -i --image=busybox:latest -- bash -c "sleep 3; date; sleep 3"
pod "test" deleted
```

Is this sufficient or would we still rather have quiet be the default with maybe the date shown?

**Comment by duglin** [comment]
Given the user asked for the pod to be removed (via the `--rm`) I would think it would be better to only show something if Kube didn't do what they ask for - meaning only print something if it didn't delete the pod. 

Not a perfect analogy, but imagine this:
```
$ rm foo
file "foo" deleted
```

Seems unnecessarily verbose to me :-)

**Comment by MHBauer** [comment]
The date needs to be shown.

**Comment by cmluciano** [comment]
@MHBauer apologies I forgot I was running the date command  in my test 😜

**Comment by cmluciano** [comment]
Can this be closed now?

**Comment by duglin** [comment]
```
$ kubectl run test --rm --restart=Never -it --image=ubuntu -- bash -c "sleep 3; date; sleep 3"
If you don't see a command prompt, try pressing enter.
                                                      Tue Feb 21 15:05:49 UTC 2017
```
I think we can remove the text about the prompt. That might be the last issue.

**Comment by ncdc** [comment]
@duglin I was tempted to avoid the prompt the last time I was in this code, but I am fairly certain I ran into some instances where the screen didn't refresh unless I pressed a key. Maybe that's not an issue any more, or I was imagining things back then 😄

**Comment by duglin** [comment]
@ncdc you may be right and a very long time ago when I was looking at the code I think there was a difference between how Kube attached to the containers vs how docker does it. I think Docker attached *before* the container started and Kube attached afterwards, thus Kube might not get all of the stdout/err stuff and why (I suspect) the prompt was added.  I was hoping we might be able to attach before the container starts like docker does... but never found the time to see if that's possible.

**Comment by ncdc** [comment]
@duglin if we wanted Kube to attach before the container started, we'd likely have to introduce some new query parameter or something to indicate that the kubelet needs to wait for the client to attach before starting the container, and there would be a whole can of worms associated with that, since you can't currently attach to a non-running pod.

We used to have the ability to run the api equivalent of `docker logs` at the beginning of the attach, prior to live-streaming the actual attach. That went away when we switched from go-dockerclient to engine-api. Docker 1.13 gives us the ability to do that again, if we want, but it's a double-edged sword. You may not want to do it 100% of the time, and you may end up streaming a *ton* of log data before you get to what's live.

**Comment by duglin** [comment]
but isn't `docker logs` just the stdout/err of the container - so what other log data are you referring to? I would think its all "live", no?

**Comment by ncdc** [comment]
Yes, `docker logs` is stdout/err of the container.

This is what we used to have:

1. start container
1. time passes (or not)
1. docker attach, logs=true (show logs, then attach)

This is what we have now:

1. start container
1. time passes (or not)
1. docker attach, logs=false (don't show logs, just attach and only show stdout/stderr emitted after attaching)

**Comment by mikebrow** [comment]
Potential help for this console pattern is working it's way from the bottom up via the CRI implementations that are based on runc. 

When using create container then start container for detached containers... with the new console (tty=true) design in runc, an option is provided on the create container command "--console-socket value" where value is a "path to an AF_UNIX socket which will receive a file descriptor referencing the master end of the console's pseudoterminal" That code was merged at the end of last year.

With this feature, after you've created the container via CRI "container create" (runc create) you should have an attached console before you call container start (runc start). 

PR underway kubernetes-incubtor/cri-o logging code PR: https://github.com/kubernetes-incubator/cri-o/pull/162

**Comment by mikebrow** [comment]
@kubernetes/sig-node

**Comment by MHBauer** [comment]
1 & 4 work as specified now.

2 still has the `Hit enter for command prompt` message and some weird spaces, but the only thing on stdout is the appropriate output. I'd be inclined to get rid of the message about the prompt.

3 is the only weird behavior where it's just different than docker.

**Comment by Doqnach** [comment]
Has case 2 ever been fixed? Because I'm running into this now...

**Comment by dperetti** [comment]
Still an issue in late 2020.

**Comment by almson** [comment]
It mostly works for me (other than #105554), but:

 - name should be auto-generated to prevent conflicts
 - text like `pod "test" deleted` should be sent to `stderr` because that's where logging messages should go. [^1]
 - the trio of `--stdin` `--tty` and `--attach` is confusing. Seems that one of those flags should die, or else the man pages should explain how they can be used in different combinations.
 - please don't kill the functionality (there's a cleanup of `kubectl run` going on--maybe a new command that is restart=Never, always attaches, and doesn't have some other irrelevant options is in order)

[^1]: As explained on [Stackoverflow](https://stackoverflow.com/a/60558096/1151521): "The purpose of the standard error stream is to separate error messages from regular output. This is important in contexts such as shell scripts, where the standard output might be sent to a pipe or to a file. That redirection leaves standard error still going to a different place — usually the terminal."
